Add support for ez80-tios

myclevorname opened the pull request:
I added 4 new integer types to Zir in order to support the eZ80's weird integers, `(u)int24_t` and `(u)int48_t`, making their usage UB (hopefully IB) outside of said target. A Hello World program using the ez80-tios target can be found at https://codeberg.org/clevor/hello-ce

Progress toward https://github.com/ziglang/zig/issues/23579

myclevorname committed:
Add support for ez80-tios

alexrp commented on lib/std/Target.zig:
```diff
@@ -676,6 +680,12 @@ pub const Os = struct {
+                        .min = .{ .major = 5, .minor = 0, .patch = 0 },
+                        .max = .{ .major = 5, .minor = 8, .patch = 3 },
+                    },
+                },
```
Where are these sourced from?

myclevorname replied:
https://education.ti.com/en/software/update/84-ce-software-update
http://datamath.org/Graphing/TI-83PRCE.htm

Wait, 5.8.4's out now? Huh.

alexrp commented on lib/std/builtin.zig:
```diff
@@ -370,6 +370,10 @@ pub const CallingConvention = union(enum(u8)) {
 
+    // Calling conventions for the `ez80` architecture.
+    ez80_sysv,
+    ez80_tiflags,
```
* Are there ABI docs for these?
* What's the difference?

myclevorname replied:
The docs for ez80_sysv are located at https://ce-programming.github.io/toolchain/static/asm.html. ez80_tiflags is ez80_sysv, but with the constraint that `IY` is 0xD00080, preserving IY in the process. This is only necessary because routines in the 83+ series rely on that register being set for both interrupts and syscalls. The CE's interrupt handler doesn't rely on IY being set anymore, but some syscalls still depend on it.

alexrp replied:
Who defined this calling convention originally - TI or the CE Toolchain people? Depending on the answer, it should be called `ez80_ti` or `ez80_ce`; we only use the `_sysv` suffix when the ABI is specified as a "processor supplement" for the generic System V ABI.

alexrp commented on lib/std/Target.zig:
```diff
@@ -740,6 +750,7 @@ pub const arm = @import("Target/arm.zig");
 pub const avr = @import("Target/avr.zig");
 pub const bpf = @import("Target/bpf.zig");
 pub const csky = @import("Target/csky.zig");
+pub const ez80 = @import("Target/ez80.zig");
```
Please just use `Target/generic.zig` until it has actual content.

alexrp commented on lib/std/Target.zig:
```diff
@@ -1134,7 +1147,6 @@ pub fn toCoffMachine(target: *const Target) std.coff.IMAGE.FILE.MACHINE {
         .x86 => .I386,
         .x86_64 => .AMD64,
 
-        .aarch64_be,
```
?

alexrp commented on lib/std/Target.zig:
```diff
@@ -1430,6 +1444,7 @@ pub const Cpu = struct {
             avr,
             bpf,
             csky,
+            ez80,
```
It's my understanding that the eZ80 is a variant of the Z80 that's binary compatible. If that is the case, the `Arch` tag should be `z80`, and any minor differences should be handled with feature bits for the eZ80.

alexrp replied:
Actually, on further reading, it seems like the eZ80 has a larger address space than the Z80? In that case, a distinct `Arch` tag is justifiable, I think. We'd want a `z80` tag for targeting the older model with a smaller address space, but that doesn't have to be part of this PR.

alexrp commented on lib/std/Target.zig:
```diff
@@ -1691,6 +1707,14 @@ pub const Cpu = struct {
+                .loongarch64,
+                .arc,
+                .propeller,
+                .ez80,
```
Looks like a failed merge? Probably same for the `aarch64_be` thing above?

alexrp commented on src/Zcu.zig:
```diff
@@ -3836,6 +3836,60 @@ pub fn atomicPtrAlignment(
+        .spirv64,
+        .loongarch64,
+        => 64,
+
```
Also looks like a failed merge.

myclevorname commented on lib/std/Target/Query.zig:
```diff
@@ -209,6 +209,7 @@ pub const ParseOptions = struct {
 };
 
 pub fn parse(args: ParseOptions) !Query {
+    @setEvalBranchQuota(21000);
```
The compiler errored about the eval branch quota getting exceeded, and the new minimum quota required in that location is 20275.

myclevorname replied:
This doesn't reproduce anymore

alexrp replied:
Even so, it's still been changed from no `@setEvalBranchQuota` call to a call with `20000`, and that still needs a justification.

myclevorname replied:
Sorry. I blame a failed rebase on my end.

alexrp commented on src/codegen/llvm.zig:
```diff
@@ -3116,6 +3119,12 @@ pub const Object = struct {
+            .i24_type,
+            .u48_type,
+            .i48_type,
+            => unreachable,
```
These are all perfectly valid Zig and LLVM types.

myclevorname committed:
ez80_sysv -> ez80_cet

alexrp commented on lib/std/Target.zig:
```diff
@@ -740,6 +750,7 @@ pub const arm = @import("Target/arm.zig");
 pub const avr = @import("Target/avr.zig");
 pub const bpf = @import("Target/bpf.zig");
 pub const csky = @import("Target/csky.zig");
+pub const ez80 = @import("Target/generic.zig");
```
```suggestion
pub const z80 = @import("Target/generic.zig");
```

This needs to match the `Cpu.Arch.Family` name.

alexrp commented on lib/std/Target.zig:
```diff
@@ -1730,13 +1748,15 @@ pub const Cpu = struct {
         /// All CPU features Zig is aware of, sorted lexicographically by name.
         pub fn allFeaturesList(arch: Arch) []const Cpu.Feature {
             return switch (arch.family()) {
+                .z80 => &Target.ez80.all_features,
```
```suggestion
```

No need for this with the suggestion above.

myclevorname committed:
Update tools/update-cpu-features.zig

myclevorname committed:
update lib/std/Target.zig

myclevorname committed:
update again

alexrp commented on lib/std/Target.zig:
```diff
@@ -1106,6 +1118,7 @@ pub fn toElfMachine(target: *const Target) std.elf.EM {
         .xcore => .XCORE,
         .xtensa, .xtensaeb => .XTENSA,
 
+        .ez80,
```
It appears to be what binutils uses for its `z80-elf` support, and binutils considers `ez80` to be a submachine of `z80`. So I think this should be fine to use here.

myclevorname committed:
set machine of ez80 to Z80

myclevorname committed:
whoops

myclevorname committed:
Use ptrAbiAlignment in more places

Should hopefully fix the alignment-related crash.

myclevorname committed:
remove

myclevorname committed:
lib/zig.h for *24, *48 partially done

myclevorname committed:
fix it

myclevorname committed:
add #

myclevorname committed:
fix it harder

myclevorname committed:
Finish the lib/zig.h intrinsics

(yay)

myclevorname committed:
fix 1

myclevorname committed:
hopefully last zig.h fix

myclevorname committed:
Fix compile error with std.Io.Writer on eZ80

myclevorname committed:
ugh

myclevorname committed:
Revert "Fix compile error with std.Io.Writer on eZ80"

This reverts commit 446c3436745dc39dddee72099f32372ea5c32208.

Dropped in favor of https://github.com/ziglang/zig/pull/25691
(0d1e26e68f94f0de76b5ab15471a34e2bb666fdd)

myclevorname committed:
Edit ez80 ptrAbiAlignment comment

alexrp commented:
Needs a rebase due to #25691 being merged.

Also I think I'd like @mlugg and @jacobly0 to review some of the compiler changes here.

alexrp commented on lib/zig.h:
```diff
@@ -3995,6 +4334,7 @@ typedef int zig_memory_order;
         return val; \
     }
 
+/* The eZ80 does not support atomics beyond 8 bits */
```
That doesn't seem relevant to MSVC atomics anyway?

myclevorname committed:
remove ez80 atomic comment

mlugg commented on lib/std/zig/llvm/Builder.zig:
```diff
@@ -234,8 +234,10 @@ pub const Type = enum(u32) {
     i1,
     i8,
     i16,
+    i24,
```
Likewise here, why add these as special cases? Again, `llvm.Builder` already has the capability to represent arbitrary integer bit widths.

mlugg commented on src/Type.zig:
```diff
@@ -1603,6 +1603,9 @@ fn abiSizeInnerOptional(
 pub fn ptrAbiAlignment(target: *const Target) Alignment {
+    // The eZ80 has 24-bit pointers, which aren't exact powers of two, tripping
+    // the assert. The alignment of eZ80 pointers is 1, so we bypass the check.
+    if (target.cpu.arch == .ez80) return .fromByteUnits(1);
```
```suggestion
    if (target.cpu.arch == .ez80) return .@"1";
```

mlugg commented on src/codegen/c.zig:
```diff
@@ -281,7 +281,9 @@ const reserved_idents = std.StaticStringMap(void).initComptime(.{
     .{ "inline", {} },
     .{ "int", {} },
     .{ "int16_t", {} },
+    .{ "int24_t", {} },
```
Ah, okay, disregard this comment then; sorry for the noise.

myclevorname committed:
last?

alexrp commented on src/Type.zig:
```diff
@@ -1603,6 +1603,9 @@ fn abiSizeInnerOptional(
 }
 
 pub fn ptrAbiAlignment(target: *const Target) Alignment {
```
Seems like this should move to `target.zig` since it has nothing to do with `Type`. But that doesn't necessarily have to happen in this PR.

myclevorname committed:
last part 2

myclevorname committed:
More portable trap I guess

The Z80 does not have "trap" instructions, but trap instructions
on the eZ80 jump to 0x000000, which is almost the same as `rst 00h`. In
addition, the Z80 and eZ80 both boot up at 0x(00)0000, so `rst 00h` is
appropriate.

andrewrk commented:
migrated to https://codeberg.org/ziglang/zig/pulls/30622
